File: scripts_python/ACSC2026Qualification_Dino_Vault.py
#!/usr/bin/env python3
"""
ACSC Qualification 2026 - Dino Vault

离线求解脚本：输入同一只恐龙（例如 Vexillum Rex）在同一连接中
两次下载得到的 (ciphertext, modulus)，利用共享素因子攻击恢复原文。

判断 RSA 的关键线索来自附件 app.py：
1. modulus = transmission_key * vault_key（两个大素数相乘）
2. exponent = 2**16 + 1 = 65537
3. ciphertext = pow(message, exponent, modulus)

这就是 textbook RSA。
"""

from __future__ import annotations

import argparse
import math
import re
import sys

from Crypto.Util.number import inverse, long_to_bytes

PUBLIC_EXPONENT = 65537
DNA_LOOKUP = {"A": 0, "T": 1, "G": 2, "C": 3}


def from_dna(dna: str) -> str:
    """把题目中的 DNA 字符串逆回普通文本。"""
    if len(dna) % 4 != 0:
        raise ValueError("DNA 长度不是 4 的倍数，无法按题目编码规则还原")

    chars: list[str] = []
    for i in range(0, len(dna), 4):
        value = 0
        for j, base in enumerate(dna[i : i + 4]):
            value |= DNA_LOOKUP[base] << (2 * j)
        chars.append(chr(value))
    return "".join(chars)


def solve(ciphertext1: str, modulus1: str, ciphertext2: str, modulus2: str) -> tuple[int, str, str | None]:
    """执行共享素因子攻击并返回 (shared_prime, plaintext, flag)。"""
    c1 = int(ciphertext1, 16)
    n1 = int(modulus1)
    n2 = int(modulus2)

    shared_prime = math.gcd(n1, n2)
    if shared_prime in (1, n1, n2):
        raise ValueError("两组模数没有得到有效共享质因子，请确认它们来自同一只恐龙的同一连接")

    q1 = n1 // shared_prime
    phi = (shared_prime - 1) * (q1 - 1)
    private_exponent = inverse(PUBLIC_EXPONENT, phi)

    message_int = pow(c1, private_exponent, n1)
    dna = long_to_bytes(message_int).decode()
    plaintext = from_dna(dna)

    match = re.search(r"dach2026\{[^}]+\}", plaintext)
    flag = match.group(0) if match else None
    return shared_prime, plaintext, flag


def main() -> None:
    try:
        sys.stdout.reconfigure(encoding="utf-8")
    except AttributeError:
        pass

    parser = argparse.ArgumentParser(description="Solve ACSC Qualification 2026 Dino Vault from two modulus/ciphertext pairs")
    parser.add_argument("--ciphertext1", required=True, help="第一组十六进制密文")
    parser.add_argument("--modulus1", required=True, help="第一组十进制模数")
    parser.add_argument("--ciphertext2", required=True, help="第二组十六进制密文")
    parser.add_argument("--modulus2", required=True, help="第二组十进制模数")
    args = parser.parse_args()

    shared_prime, plaintext, flag = solve(args.ciphertext1, args.modulus1, args.ciphertext2, args.modulus2)

    print(f"[+] shared prime bits: {shared_prime.bit_length()}")
    print(f"[+] shared prime: {shared_prime}")
    print(f"[+] recovered plaintext: {plaintext}")
    if flag:
        print(f"[+] flag: {flag}")
    else:
        print("[!] 未在明文中自动匹配到 flag，请手动检查 recovered plaintext")


if __name__ == "__main__":
    main()

File: scripts_go/ACSC2026Qualification_Dino_Vault.go
package main

import (
	"flag"
	"fmt"
	"math/big"
	"regexp"
)

// ACSC Qualification 2026 - Dino Vault
//
// 离线求解脚本：输入同一只恐龙（例如 Vexillum Rex）在同一连接中
// 两次下载得到的密文与模数，利用共享素因子攻击恢复明文。
//
// 关键判断 RSA 的依据：
//   1. n = p * q（两个大素数相乘）
//   2. e = 2^16 + 1 = 65537
//   3. c = m^e mod n
//
// 运行示例：
// go run CTF_Writeups/scripts_go/ACSC2026Qualification_Dino_Vault.go \
//   -ciphertext1 <hex1> -modulus1 <n1> \
//   -ciphertext2 <hex2> -modulus2 <n2>

const publicExponent int64 = 65537

var dnaLookup = map[byte]int64{
	'A': 0,
	'T': 1,
	'G': 2,
	'C': 3,
}

func mustParseHex(name, raw string) *big.Int {
	value, ok := new(big.Int).SetString(raw, 16)
	if !ok {
		panic(fmt.Sprintf("invalid hex for %s", name))
	}
	return value
}

func mustParseDec(name, raw string) *big.Int {
	value, ok := new(big.Int).SetString(raw, 10)
	if !ok {
		panic(fmt.Sprintf("invalid decimal for %s", name))
	}
	return value
}

func mustInverseModulo(e, phi *big.Int) *big.Int {
	inverse := new(big.Int).ModInverse(e, phi)
	if inverse == nil {
		panic("modular inverse does not exist")
	}
	return inverse
}

func fromDNA(dna string) string {
	if len(dna)%4 != 0 {
		panic("dna length is not a multiple of 4")
	}

	plaintext := make([]byte, 0, len(dna)/4)
	for i := 0; i < len(dna); i += 4 {
		var value int64
		for j := 0; j < 4; j++ {
			base := dna[i+j]
			decoded, ok := dnaLookup[base]
			if !ok {
				panic(fmt.Sprintf("unexpected DNA base %q", base))
			}
			value |= decoded << (2 * j)
		}
		plaintext = append(plaintext, byte(value))
	}
	return string(plaintext)
}

func extractFlag(plaintext string) string {
	re := regexp.MustCompile(`dach2026\{[^}]+\}`)
	return re.FindString(plaintext)
}

func main() {
	ciphertext1Raw := flag.String("ciphertext1", "", "first ciphertext in hex")
	modulus1Raw := flag.String("modulus1", "", "first modulus in decimal")
	ciphertext2Raw := flag.String("ciphertext2", "", "second ciphertext in hex")
	modulus2Raw := flag.String("modulus2", "", "second modulus in decimal")
	flag.Parse()

	if *ciphertext1Raw == "" || *modulus1Raw == "" || *ciphertext2Raw == "" || *modulus2Raw == "" {
		flag.Usage()
		panic("all four inputs are required")
	}

	c1 := mustParseHex("ciphertext1", *ciphertext1Raw)
	n1 := mustParseDec("modulus1", *modulus1Raw)
	n2 := mustParseDec("modulus2", *modulus2Raw)

	sharedPrime := new(big.Int).GCD(nil, nil, n1, n2)
	if sharedPrime.Cmp(big.NewInt(1)) == 0 || sharedPrime.Cmp(n1) == 0 || sharedPrime.Cmp(n2) == 0 {
		panic("no useful shared prime found; ensure both pairs come from the same dino in one connection")
	}

	q1 := new(big.Int).Quo(new(big.Int).Set(n1), sharedPrime)
	phi := new(big.Int).Mul(
		new(big.Int).Sub(new(big.Int).Set(sharedPrime), big.NewInt(1)),
		new(big.Int).Sub(new(big.Int).Set(q1), big.NewInt(1)),
	)

	e := big.NewInt(publicExponent)
	d := mustInverseModulo(e, phi)

	messageInt := new(big.Int).Exp(c1, d, n1)
	dna := string(messageInt.Bytes())
	plaintext := fromDNA(dna)
	flagValue := extractFlag(plaintext)

	fmt.Printf("[+] shared prime bits: %d\n", sharedPrime.BitLen())
	fmt.Printf("[+] shared prime: %s\n", sharedPrime.String())
	fmt.Printf("[+] recovered plaintext: %s\n", plaintext)
	if flagValue != "" {
		fmt.Printf("[+] flag: %s\n", flagValue)
	} else {
		fmt.Println("[!] no dach2026{...} flag pattern found; inspect recovered plaintext manually")
	}
}
